spring-boot-modules/spring-boot-keycloak/spring-boot-mvc-client/src/main/java/com/baeldung/boot/keycloak/client/SecurityConfig.java

```diff
@@ -0,0 +1,76 @@
+package com.baeldung.boot.keycloak.client;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.http.HttpMethod;
+import org.springframework.security.config.Customizer;
+import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.authority.mapping.GrantedAuthoritiesMapper;
+import org.springframework.security.oauth2.client.oidc.web.logout.OidcClientInitiatedLogoutSuccessHandler;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+import org.springframework.security.oauth2.core.oidc.OidcIdToken;
+import org.springframework.security.oauth2.core.oidc.user.OidcUserAuthority;
+import org.springframework.security.web.SecurityFilterChain;
+
+@Configuration
+@EnableWebSecurity
+@EnableMethodSecurity
+public class SecurityConfig {
+
+  static interface AuthoritiesConverter
+      extends Converter<Map<String, Object>, Collection<GrantedAuthority>> {
+  }
+
+  @Bean
+  AuthoritiesConverter realmRolesAuthoritiesConverter() {
+    return claims -> {
+      final var realmAccess = Optional.ofNullable((Map<String, Object>) claims.get("realm_access"));
+      final var roles =
+          realmAccess.flatMap(map -> Optional.ofNullable((List<String>) map.get("roles")));
+      return roles.map(List::stream).orElse(Stream.empty()).map(SimpleGrantedAuthority::new)
+          .map(GrantedAuthority.class::cast).toList();
```

Comment:
I feel like this could be refactored into a single fluent statement:

```java
return claims -> Optional.ofNullable((Map<String, Object>) claims.get("realm_access"))
  .flatMap(map -> Optional.ofNullable((List<String>) map.get("roles")))
  .map(List::stream)
  .orElse(Stream.empty())
  .map(SimpleGrantedAuthority::new)
  .map(GrantedAuthority.class::cast)
  .toList();
```

And then I'd wonder if this cannot be simplified. Using `Optional::stream` for example.

Reply:
> this could be refactored into a single fluent statement

Yes, it can. However, the usage of intermediate named references has no impact on performance (I do not collect a stream for instance), and it makes:
- code clearer for beginners: at this point, you have `realm` claims, at this one you extract the roles claim from realm ones, etc.
- step by step debugging easier (for instance for someone wanting to switch from realm roles to client ones).

> I'd wonder if this cannot be simplified. Using `Optional::stream` for example

Yes, it can be refactored using `Optional::stream`:
```java
return roles
  .stream()
  .flatMap(List::stream)
  .map(SimpleGrantedAuthority::new)
  .map(GrantedAuthority.class::cast).toList();
```
However:
- this does not save a statement (`.stream().flatMap(List::stream)` VS `.map(List::stream).orElse(Stream.empty())`
- my experience training beginners & intermediate Java developers taught me that the usage of `flatMap` has a significant impact on audience attention: when I use it, a surprising amount of attendees stop following me.

So unless you explicitly instruct me otherwise, I'll keep this code as is.